sentinel1_reader/sentinel1_reader.py

```diff
@@ -0,0 +1,387 @@
+        Sensing start of burst; taken from azimuth time
+    sensing_stop : datetime.datetime
+        Sensing stop of burst
+    osv_list : xml.etree.ElementTree.Element
+        ElementTree containing orbit state vectors
+
+    Returns:
+    --------
+    _ : datetime
+        Sensing mid as datetime object.
+    '''
+    fmt = "UTC=%Y-%m-%dT%H:%M:%S.%f"
+    orbit_sv = []
+    # add start & end padding to ensure sufficient number of orbit points
+    pad = datetime.timedelta(seconds=60)
+    for osv in osv_list:
+        t_orbit = datetime.datetime.strptime(osv[1].text, fmt)
+        pos = [float(osv[i].text) for i in range(4,7)]
+        vel = [float(osv[i].text) for i in range(7,10)]
+        if t_orbit > sensing_start - pad:
+            orbit_sv.append(isce3.core.StateVector(isce3.core.DateTime(t_orbit),
+                                                   pos, vel))
+        if t_orbit > sensing_stop + pad:
+            break
+
+    # use list of stateVectors to init and return isce3.core.Orbit
+    time_delta = datetime.timedelta(days=2)
+    ref_epoch = isce3.core.DateTime(sensing_start - time_delta)
+
+    return isce3.core.Orbit(orbit_sv, ref_epoch)
+
+def get_burst_centers_and_boundaries(tree):
+    '''
+    Parse grid points list and calculate burst center lat and lon
+
+    Parameters:
+    -----------
+    tree : Element
+        Element containing geolocation grid points.
+
+    Returns:
+    --------
+    _ : list
+        List of burst centroids ass shapely Points
+    _ : list
+        List of burst boundaries as shapely Polygons
+    '''
+    # find element tree
+    grid_pt_list = tree.find('geolocationGrid/geolocationGridPointList')
+
+    # read in all points
+    n_grid_pts = int(grid_pt_list.attrib['count'])
+    lines = np.empty(n_grid_pts)
+    pixels = np.empty(n_grid_pts)
+    lats = np.empty(n_grid_pts)
+    lons = np.empty(n_grid_pts)
+    for i, grid_pt in enumerate(grid_pt_list):
+        lines[i] = int(grid_pt[2].text)
+        pixels[i] = int(grid_pt[3].text)
+        lats[i] = float(grid_pt[4].text)
+        lons[i] = float(grid_pt[5].text)
+
+    unique_line_indices = np.unique(lines)
+    n_bursts = len(unique_line_indices) - 1
+    center_pts = [[]] * n_bursts
+    boundary_pts = [[]] * n_bursts
+
+    # zip lines numbers of bursts together and iterate
+    for i, (ln0, ln1) in enumerate(zip(unique_line_indices[:-1],
+                                       unique_line_indices[1:])):
+        # create masks for lines in current burst
+        mask0 = lines==ln0
+        mask1 = lines==ln1
+
+        # reverse order of 2nd set of points so plots of boundaries
+        # are not connected by a diagonal line
+        burst_lons = np.concatenate((lons[mask0], lons[mask1][::-1]))
+        burst_lats = np.concatenate((lats[mask0], lats[mask1][::-1]))
+
+        poly = shapely.geometry.Polygon(zip(burst_lons, burst_lats))
+        center_pts[i] = poly.centroid.xy
+        boundary_pts[i] = check_dateline(poly)
```

Comment:
Shouldn't we apply the `dateline` checking before calculating the centroid? One simple way could be just to use the largest polygons among the returned list of polygons.

Reply:
Maybe @vbrancat or @hfattahi knows better, I would think we want only one `center_pts` for each burst, so we only need the largest polygon, instead of all the spited ones.

Reply:
Why would the center of the largest split poly be preferred to center of the burst? IIRC polygon splitting was to accommodate DEM downloading?

BTW both `Polygon.centroid` and the previous averaging of points do a poor job of calculating the actual center of the burst when its split by the dateline. AFAICT using `Polygon.centroid` as-is gives wrong results. The averaging method is a bit off. The following seems to work better:

1. convert LLH to XYZ
2. average XYZ
3. covert average XYZ to LLH
4. set average LLH height to 0

Reply:
Interesting. I think it should work!

Reply:
I like the XYZ calculation too.